Route documents to the correct shards in tsdb

nik9000 opened the pull request:
This causes elasticsearch to land documents from the same time series on
the same shard. It does so by adding a new index setting `routing_path`
which must be set when an index is in `mode: time_series` and may not be
set outside of that mode. That setting contains a list of patterns to
extract from the `_source` document that are hashed into the routing
value.

Note: This doesn't guarantee that the `routing_path` only matches dimensions.
For now it is possible to configure the `routing_path` so it matches non-dimension
fields. This would the same time series to end up in multiple shards. We don't want
that and plan to add that constraint in a follow up change.

nik9000 committed:
Route documents to the correct shards in tsdb

This causes elasticsearch to land documents from the same time series on
the same shard. It does so by adding a new index setting `routing_path`
which must be set when an index is in `mode: time_series` and may not be
set outside of that mode. That setting contains a list of patterns to
extract from the `_source` document that are hashed into the routing
value.

nik9000 commented:
In addition to routing documents to the correct shard this adds tsdb's restriction turning off delete and update and indexing with a specific routing or id.

nik9000 committed:
Merge branch 'master' into index_routing_from_source

imotov commented on qa/smoke-test-multinode/src/test/resources/rest-api-spec/test/smoke_test_multinode/20_tsdb_consistency.yml:
```diff
@@ -0,0 +1,103 @@
+# Test the mode:time_series properly groups by _tsid. If we could put this in
+# rest-api-spec we would, but it requires painless.
+
```
Should we skip setup here as well? I am curious why this didn't fail in bwc test?

nik9000 replied:
Oh! It's because we don't run this against a mixed version cluster! Huh. I'll have a think about how to do that.

imotov replied:
I think tests were also disabled, because it now fails in https://github.com/elastic/elasticsearch/pull/77731/files#diff-ee59a8405494ffd00e1b28d5f3ed8d537b51fa846deed022fac7fc42328fc698R13

imotov commented on build-tools-internal/src/main/groovy/elasticsearch.formatting.gradle:
```diff
@@ -207,6 +207,7 @@ subprojects {
                    'src/*/java/org/elasticsearch/action/admin/cluster/snapshots/**/*.java',
                    'src/test/java/org/elasticsearch/common/xcontent/support/AbstractFilteringTestCase.java',
                    'src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java',
+                   'src/*/java/org/elasticsearch/index/IndexRouting.java',
```
Nice! 😄

imotov reviewed (commented):
Looks good to me from TSDB perspective. It doesn't have everything we need, but it is a good start.

nik9000 committed:
Merge branch 'master' into index_routing_from_source

nik9000 committed:
Merge branch 'master' into index_routing_from_source

nik9000 committed:
Moar skip

nik9000 committed:
tsdb survives full cluster restart

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java:
```diff
@@ -116,4 +205,158 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
+            XContentType sourceType,
+            BytesReference source
+        ) {
+            if (idIsAutoGenerated == false) {
+                throw new IllegalArgumentException(error("indexing with a specified id"));
+            }
```
This seems like more of a timeseries db specific request validation, I see no reason to enforce auto-id otherwise? I wonder if it is a bit premature to add it here even, since the logic to build the `id` is not yet added. I think `IndexRequest.process` will need some kind of indication anyway that it is a timeseries db request and thus it may move there in the end? The only use of this variable is in this validation, so if we remove it here, the interface for `indexShard` will become simpler too.

nik9000 replied:
I can drop it for now but we need to do *something* about this before we remove the feature flag. Our hope was to disable auto-generated ID stuff for timeseries so we could deal with the all the plumbing of not generating the id at all at a later stage. But I can pull this out and we can keep it on the TODO list.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java:
```diff
@@ -116,4 +205,158 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
+            int firstHash = extractItem(source);
+            if (source.currentToken() == Token.END_OBJECT) {
+                // Just one routing key in this object
+                // Use ^ like Map.Entery's hashcode
```
```suggestion
                // Use ^ like Map.Entry's hashcode
```

nik9000 replied:
:+1:

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java:
```diff
@@ -116,4 +205,158 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
+                // Use ^ like Map.Entery's hashcode
+                return Murmur3HashFunction.hash(firstFieldName) ^ firstHash;
+            }
+            String[] fields = new String[] { firstFieldName, null };
+            int[] hashes = new int[] { firstHash, 0 };
```
I wonder if this is better than just having a `List<FieldAndHash> fields = new ArrayList<>()`? AFAIK, we will often have more than 2 fields. And allocating an object per (field,hash) pair seems unproblematic, we are parsing json (and indexing) after all here.

nik9000 replied:
Here's the `routing_path` that I put together for some k8s data:
```
    "routing_path": [
      "kubernetes.container.id",
      "kubernetes.event.involved_object.uid",
      "kubernetes.node.name",
      "kubernetes.pod.name",
      "kubernetes.system.container",
      "kubernetes.volume.name",
      "metricset.name"
    ]
```

In my data there are always `metricset.name` and at least one of the others.

I think it'll be quite common for a "leaf" to have a single value. So I like the single-value optimization. But I think you are right about the `List<FieldAndHash>` for the more than one value case.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java:
```diff
@@ -116,4 +205,158 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
+                String fieldName = source.currentName();
+                source.nextToken();
+                if (length >= fields.length) {
+                    fields = ArrayUtil.grow(fields, length + 1);
```
Would we not want to grow by more than 1? This seems expensive for the cases where we have 5-10 fields that match.

nik9000 replied:
`length+1` is a minimum here. `grow` will oversize the array based on some heuristics about the jvm.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java:
```diff
@@ -116,4 +205,158 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
+                            NamedXContentRegistry.EMPTY,
+                            DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                            source.streamInput(),
+                            include,
```
Do you have plans to optimize this for the 10K+ filter case? I did not dig into whether we can be smarter on the xcontent parsing/filter-path here, but current implementation looks somewhat expensive for that case.

nik9000 replied:
Yes indeed. I have plans to do more benchmarking on the case where there are a couple thousand of these filters. The `FilterPathBasedFilter` looks linear the number of filters at any particular level. I think it'd be simple enough to use a HashMap there, but I haven't looked hard yet.

That 30k number I've been throwing around is a gross estimate of the length of the text that makes up the filter.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java:
```diff
@@ -116,4 +205,158 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
+                throw new IllegalArgumentException(error("indexing with a specified id"));
+            }
+            if (routing != null) {
+                throw new IllegalArgumentException(error("indexing with a specified routing"));
```
I think we would want a GET request to use routing. That is obviously irrelevant for this method. But I wonder if we would then not also support an indexing request with the "right" routing value. That would allow updates and deletes too, meaning update-by-query and delete-by-query would work?

So that makes me think whether we should just calculate a routing value rather than do the hashing to shard-id here?

nik9000 replied:
The old PR calculated a routing value and that introduces a ton of complexity around what happens with it in the translog. We didn't like the idea of writing this generated routing value to the translog because its just duplicate bytes. It felt much more natural not to have to look at `routing` at all for tsdb and make the change we needed here. At least it did to me.

For what it's worth tsdb doesn't plan to support get and update and delete at all because we don't plan to have IDs at all. We have them now. And we'll probably have them in the first release. But we'd like to remove them.

This is one of the reasons why I wanted to tie this routing generation to the tsdb project - what you are proposing sound much more complex than what I did and tsdb doesn't want it.

nik9000 replied:
@henningandersen and I talked for a while about this point. We *think* it should be possible to implement stuff like `GET` and `UPDATE` based on this approach. Like I said, TSDB doesn't need them now. But its nice that we think its possible to get them from this one day if we need them badly enough.

The short version is that if we decide to use `_tsid:@timestamp` as the `_id` we can decode the routing from it and use that for `GET` and probably even `UPDATE` and update-by-query and delete-by-query.

We talked about an alternate approach that builds a `_routing` value from this information rather than a hash. Either we build it on the fly all the time or we build it and store it. Storing it is not *terrible* because it'll compress well with `_source` in stored fields. But we'd pay full price for it in the translog and over the wire. Generating on the fly is fairly wasteful compared to generating the hash. It's not a terrible idea, but it we think it'd be more complex than generating an `_id` on the fly. So that's one way we could get UPDATE and friends from this: generate the `_id` as `_tsid:@timestamp`. We're not sure we want to do that. But just having *a* way to do all this validates this approach.

henningandersen replied:
We discussed this on another channel and we think a better approach than storing the routing is to generate a `_id` on search and then extract the routing information necessary on get, index, update, delete. While there is a lot to be untangled to get there (and we may choose never to do so), it does mean that a door is open for supporting those requests. Which in turn means that the approach taken in this PR looks good to me.

henningandersen reviewed (commented):
Thanks Nik, I left a few comments. I did not find a restriction on GET in the design doc, hence the question on that, we may need to sync up on that.

nik9000 commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -461,29 +467,35 @@ protected void doRun() {
                             MappingMetadata mappingMd = indexMetadata.mapping();
                             Version indexCreated = indexMetadata.getCreationVersion();
                             indexRequest.resolveRouting(metadata);
+                            boolean idIsAutoGenerated = indexRequest.id() == null;
```
Yeah - I was looking at that but didn't do it because I didn't like that there were two ways to know depending on where you are in the process. If you like the method I can add it though. It makes sense. Its just funny there are two ways to tell.

nik9000 committed:
Remove auto generated id rejection

We do want to reject these documents but let's sae that for a follow up
change.

nik9000 committed:
Simplify

nik9000 commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -328,6 +328,14 @@ public static APIBlock readFrom(StreamInput input) throws IOException {
     public static final Setting<Integer> INDEX_FORMAT_SETTING =
             Setting.intSetting(INDEX_FORMAT, 0, Setting.Property.IndexScope, Setting.Property.Final);
 
+    public static final Setting<List<String>> INDEX_ROUTING_PATH = Setting.listSetting(
```
In another sense, when `index.routing_path` is set routing is *forbidden* on CRUD operations. And I think that sense aligns better with how we document it. And how the code works, thankfully.

nik9000 replied:
[Here](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-routing-field.html#_making_a_routing_value_required) are the docs we have for this.

nik9000 committed:
Forbid routing_required

nik9000 committed:
Merge branch 'master' into index_routing_from_source

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

henningandersen commented on qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java:
```diff
@@ -210,6 +222,124 @@ public void testNewReplicasWork() throws Exception {
+            .entry(
+                "hits",
+                matchesMap().extraOk().entry("total", Map.of("value", numDocs, "relation", "eq")))
+            .entry("aggregations", Map.of("check", Map.of("value", IntStream.range(0, 10).mapToObj(i -> "dim" + i).collect(toList()))))
```
Aggs question: are we sure of the order of the returned values in the values list here?

nik9000 replied:
The script uses a TreeSet to get that ordering.

henningandersen commented on qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java:
```diff
@@ -210,6 +222,124 @@ public void testNewReplicasWork() throws Exception {
+        if (isRunningAgainstOldCluster()) {
+            createTimeSeriesModeIndex(0);
+        } else {
+            final int numReplicas = 1;
```
Perhaps add a comment here that we rely on 2 data nodes in the cluster?

nik9000 replied:
Sure.we.tun these tests with 3 or 4 nodes and I'll bet most.of them rely on it.

nik9000 replied:
I was mixed up - this one runs with two nodes. The rolling upgrade runs with three.

henningandersen reviewed (commented):
Thanks Nik, I left a few comments. The main concern is about forking from the bulk action when the request contains tsdb items.

nik9000 commented on qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java:
```diff
@@ -210,6 +222,124 @@ public void testNewReplicasWork() throws Exception {
     }
 
+    public void testSearchTimeSeriesMode() throws Exception {
+        assumeTrue("time series mode introduced in 8.0.0 to be backported to 7.15.0", getOldClusterVersion().onOrAfter(Version.V_8_0_0));
```
:+1:. Copy and paste mistake from the prototype.

nik9000 commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -462,28 +468,32 @@ protected void doRun() {
+                                docWriteRequest.id(),
+                                docWriteRequest.routing(),
+                                indexRequest.getContentType(),
+                                indexRequest.source()
```
I'll have a look!

nik9000 commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -462,28 +468,32 @@ protected void doRun() {
-                        concreteIndex,
-                        idx -> IndexRouting.fromIndexMetadata(clusterState.metadata().getIndexSafe(idx))
+                    List<BulkItemRequest> shardRequests = requestsByShard.computeIfAbsent(
+                        RoutingTable.shardRoutingTable(clusterState.routingTable().index(concreteIndex), shardId).getShardId(),
```
:+1:. This came from moving code out of OperationRouting. I'm still not really familiar with these classes yest. By the time I'm all done here maybe.

nik9000 commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -462,28 +468,32 @@ protected void doRun() {
                             Version indexCreated = indexMetadata.getCreationVersion();
                             indexRequest.resolveRouting(metadata);
                             indexRequest.process(indexCreated, mappingMd, concreteIndex.getName());
+                            shardId = indexRouting.indexShard(
```
I thought about doing that when I was out on a walk a while back. I didn't like it at the time. But I'll give it a shot and see how it looks.

nik9000 committed:
Merge branch 'master' into index_routing_from_source

nik9000 committed:
Small

nik9000 committed:
Fork fork knife

nik9000 commented:
@henningandersen I pushed a change that adds the forking. I'd love it if you could see if it looks like it is along the right lines. I haven't had a chance to write new tests for it, but the old tests seem happy enough.

nik9000 commented:
> old tests seem happy enough.

Not quite! DataStreams tests failed. I'll have a look on Monday.

nik9000 committed:
Let failures flow

nik9000 commented:
> Not quite! DataStreams tests failed. I'll have a look on Monday.

I think I tracked this down. This is ready for a look to see if the way I'm forking makes sense. I haven't added new tests around it, but would love opinions on the direction.

nik9000 committed:
Merge branch 'master' into index_routing_from_source

nik9000 committed:
Fix full cluster

nik9000 commented:
> I think I tracked this down. This is ready for a look to see if the way I'm forking makes sense. I haven't added new tests around it, but would love opinions on the direction.

I'm fixing some build failures from merging in master. Those should affect the approach though.

nik9000 committed:
Merge branch 'master' into index_routing_from_source

nik9000 committed:
Always fork

nik9000 commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -394,37 +406,46 @@ private long buildTookInMillis(long startTimeNanos) {
      * retries on retryable cluster blocks, resolves item requests,
      * constructs shard bulk requests and delegates execution to shard bulk action
      * */
-    private final class BulkOperation extends ActionRunnable<BulkResponse> {
```
Removing `ActionRunnable` here felt more important when we optionally forked. Now that we've mostly removed that I don't think I *need* to remove this. But I think it is more clear without it.

nik9000 commented:
I think *always* forking has exposed some fun bugs that we may have in other places. I'm not sure! See my change to `Retry`. Also, `IndexingPressureIT` fails about half the time after always forking. I'll bet it has assumptions about not forking. I wonder if I can get help on that one.

nik9000 committed:
Retry?

nik9000 committed:
Remove pressure test arm

nik9000 committed:
Merge branch 'master' into index_routing_from_source

nik9000 committed:
Add missing settings

nik9000 committed:
WIP

henningandersen commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -824,6 +884,23 @@ synchronized void markItemAsFailed(int slot, Exception e) {
         }
+    }
 
+    boolean routingMustForkFromTransportThread(Iterable<DocWriteRequest<?>> requests, ConcreteIndices concreteIndices) {
```
I think this is now unused?

henningandersen reviewed (commented):
> This is ready for a look to see if the way I'm forking makes sense
Yes, I think the way the forking is done now makes sense.

nik9000 committed:
Merge branch 'master' into index_routing_from_source

nik9000 committed:
Remove leftover

nik9000 committed:
More cleaning

nik9000 committed:
Fixup more tests

nik9000 committed:
Remove old skip

nik9000 committed:
New tests for Retry

nik9000 committed:
More tests

nik9000 committed:
Revert unrelated

nik9000 committed:
Merge branch 'master' into index_routing_from_source

henningandersen reviewed (approved):
LGTM, though I think it makes sense to get a review from a tsdb team member too.

imotov reviewed (approved):
LGTM

nik9000 committed:
Merge branch 'master' into index_routing_from_source

nik9000 committed:
One dispatch please

nik9000 committed:
Stuff moved

nik9000 committed:
More moving

nik9000 committed:
Explain why fork

nik9000 committed:
Back to ActionRunnable

nik9000 committed:
Update comment

nik9000 committed:
Utility method

nik9000 committed:
Move routing_path under feature flag

nik9000 committed:
Imports

nik9000 committed:
Merge branch 'master' into index_routing_from_source

nik9000 commented:
@elasticmachine test this please

nik9000 commented:
Tests are passing, but I want to run them a few more times out of paranoia.

nik9000 commented:
@elasticmachine test this please
